02.calendar/cal.js

```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env node
+
+import minimist from "minimist";
+import dayjs from "dayjs";
+import ja from "dayjs/locale/ja.js";
+
+dayjs.locale(ja);
+
+const createSpace = (dayNumber) => {
+  return "   ".repeat(dayNumber);
+};
+
+const showCalendar = () => {
+  const firstDate = dayjs(`${year}-${month}-1`);
+  const lastDay = firstDate.endOf("M").format("D");
+
+  for (let day = 1; day <= lastDay; day++) {
+    const currentDate = firstDate.add(day - 1, "d");
+
+    if (day === 1) {
+      const dayNumber = currentDate.format("d");
+      process.stdout.write(createSpace(dayNumber));
+    }
+    process.stdout.write(`${currentDate.format("D").padStart(2)} `);
+
+    const saturday = "6";
+    if (currentDate.format("d") === saturday) {
+      console.log("");
+    }
+  }
+};
+
+const argv = minimist(process.argv.slice(2));
+
+const today = dayjs();
+const month = argv["m"] ? String(argv["m"]) : today.format("M");
+const year = argv["m"] && argv["y"] ? String(argv["y"]) : today.format("YYYY");
+
+console.log(`      ${month}月 ${year}`);
+console.log("日 月 火 水 木 金 土");
+
+showCalendar();
```

Comment:
```suggestion
showCalendar(year, month);
```

argvの処理だけトップレベルで実装し、それ以外はshowCalendarに集約、トップレベル変数（yearとmonth）への依存もなくしてshowCalendarメソッドの再利用性を向上させる、みたいなアプローチにするとさらにいいかも

Reply:
伊藤さん、ご指摘ありがとうございます！
意図を正しく汲み取れているか不安ですが、修正できました👨‍💻

> showCalendarに集約、トップレベル変数（yearとmonth）への依存もなくしてshowCalendarメソッドの再利用性を向上させる

そういう視点もありませんでした...
まだまだプロの道は長いですね！コツコツ頑張ります💪

Reply:
あ、残念。ちょっと僕の意図と違いますね😅
前回のコメントで、

```javascript
showCalendar(year, month);
```

って書いてたので、

```javascript
const today = dayjs();
const month = argv["m"] ? String(argv["m"]) : today.format("M");
const year =
  argv["m"] && argv["y"] ? String(argv["y"]) : today.format("YYYY");
showCalendar(year, month);
```

としましょう、ということが言いたかったのでした。

こうしておけば、

```javascript
showCalendar(2021, 11);
showCalendar(1977, 7);
showCalendar(2400, 1);
```

みたいに引数で年月が自由に指定できます。
もちろん今の実装（というか一つ前の実装？）でも

```javascript
const year = 2022;
const month = 11;
showCalendar();
```

みたいにすれば同じようなことはできますが、このコードだけ見てもyearとmonthがshowCalendarで使われることは読み解けないので、`showCalendar(2021, 11);` と書けた方が自然でわかりやすいはずです。

あとRubyでもJSでもそうですが、 argv のような起動時引数は特殊な値なので扱いが難しいです。
最新のコードでは

```javascript
function showCalendar() {
  const today = dayjs();
  const month = argv["m"] ? String(argv["m"]) : today.format("M");
```

となっていますが、 argv の値はプログラムが実行された時点で決まるので、showCalendar内では変更のしようがありません。

一方、

```javascript
function showCalendar(year, month) {
```

になっていれば、argvと違ってyearとmonthはこの関数を呼び出す側で自由に決められるので、プログラムの柔軟性が向上します。

もちろん、この課題では「yearとmonthを自由に変更したい」という要件はないので、そこまで考えなくて良い、という見方もできますが、対応のコストが大きくないのなら、汎用性が高く、柔軟な設計にしておくに越したことはありません。

Reply:
こんなに丁寧に...ありがとうございます！
また、書いてあったのに意図を汲み取れておらず申し訳ございませんでした。🙏

> 対応のコストが大きくないのなら、汎用性が高く、柔軟な設計にしておくに越したことはありません。

僕も経験を積んで、自分で考えて設計していけるようになります！
教えていただきありがとうございました🙇‍♂️